crates/solver/src/settlement_submission.rs

```diff
@@ -9,23 +9,94 @@ use anyhow::{anyhow, Result};
 use contracts::GPv2Settlement;
 use ethcontract::{
     errors::{ExecutionError, MethodError},
-    Account, TransactionHash,
+    Account, Address, TransactionHash,
 };
 use futures::FutureExt;
-use gas_estimation::GasPriceEstimating;
+use gas_estimation::{EstimatedGasPrice, GasPriceEstimating};
 use primitive_types::{H256, U256};
 use shared::Web3;
 use std::{
-    sync::Arc,
+    collections::HashMap,
+    sync::{Arc, Mutex},
     time::{Duration, Instant},
 };
 use submitter::{
-    DisabledReason, Submitter, SubmitterGasPriceEstimator, SubmitterParams, TransactionSubmitting,
+    DisabledReason, Strategy, Submitter, SubmitterGasPriceEstimator, SubmitterParams,
+    TransactionHandle, TransactionSubmitting,
 };
 use web3::types::TransactionReceipt;
 
 const ESTIMATE_GAS_LIMIT_FACTOR: f64 = 1.2;
 
+// Key (Address, U256) represents pair (sender, nonce)
+type SubTxPool = HashMap<(Address, U256), Vec<(TransactionHandle, EstimatedGasPrice)>>;
+
+#[derive(Default)]
+struct TxPool {
+    eden: SubTxPool,
+    flashbots: SubTxPool,
+    custom_nodes: SubTxPool,
+}
+
+impl TxPool {
+    pub fn get_sub_pool(&self, strategy: Strategy) -> &SubTxPool {
+        match strategy {
+            Strategy::Eden => &self.eden,
+            Strategy::Flashbots => &self.flashbots,
+            Strategy::CustomNodes => &self.custom_nodes,
+        }
+    }
+
+    pub fn get_sub_pool_mut(&mut self, strategy: Strategy) -> &mut SubTxPool {
+        match strategy {
+            Strategy::Eden => &mut self.eden,
+            Strategy::Flashbots => &mut self.flashbots,
+            Strategy::CustomNodes => &mut self.custom_nodes,
+        }
+    }
```

Comment:
Seems like it causes a lot of hassle that the `TxPool` knows about all the different sub pools.
Would it make things easier if there is a unique `SubTxPool` associated with each submission strategy?
From a first glance this could be part of the `TransactionStrategyArg` although I have to admit this piece of data "feels" different compared to the other fields in `TransactionStrategyArg`.
I'm not yet sure how I feel about this idea myself but I wanted to mention it in case you like it.

Reply:
I thought of reusing the `TransactionStrategy` enum which is strictly used for settlement submission, however then I would need to make enum variant values (`StrategyArgs`) optional, and also, wouldn't know what to do with `DryRun` as a special case, so after all I felt like introducing a new enum is far easier... Yet, I don't have a strong opinion here what is better...

If I understand the first part of the comment, you would propose to substitute 
```
pub fn get_sub_pool(&self, strategy: Strategy) -> &SubTxPool {
        match strategy {
            Strategy::Eden => &self.eden,
            Strategy::Flashbots => &self.flashbots,
            Strategy::CustomNodes => &self.custom_nodes,
        }
    }
```

with 
```
pub fn get_eden_pool(&self) -> &SubTxPool {
   &self.eden,
}
pub fn get_flashbots_pool(&self) -> &SubTxPool {
   &self.flashbots,
}
pub fn get_custom_nodes_pool(&self) -> &SubTxPool {
   &self.custom_nodes,
}
```

Reply:
Not quite. Since every `Submitter` only cares about the tx pool of the particular strategy (right?) I wanted to make `SubTxPool` part of the `StrategyArgs`. When the `SolutionSubmitter` creates a `Submitter` it stores only the tx pool belonging to the strategy in it.
When then the `Submitter` needs to access the right tx pool it no longer has to know which specific `SubTxPool` to request because there is only one (the one the `SolutionSubmitter` stored in the `Submitter`).

The only important code changes would be:
```rust
pub struct StrategyArgs {
    pub submit_api: Box<dyn TransactionSubmitting>,
    pub max_additional_tip: f64,
    pub additional_tip_percentage_of_max_fee: f64,
    // add tx pool which only belongs to this strategy
    pub transaction_pool: Arc<Mutex<SubTxPool>>,
}

// store correct tx pool when SolutionSubmitter::settle creates each Submitter
let submitter = Submitter::new(
    &self.contract,
    &account,
    strategy_args.submit_api.as_ref(),
    &gas_price_estimator,
    self.access_list_estimator.as_ref(),
    strategy_args.transaction_pool.clone()
)?;
```
The rest should follow from fixing the caused type errors.

Reply:
I understand now. So, yes you are right, each `Submitter` cares about it's own tx pool (for now), however, I aimed to leave a bit more flexibility in case we need each `Submitter` to also take an occasional look at other subpools to resolve some strange bugs like: https://github.com/cowprotocol/services/issues/204

Reply:
The issue with hardcoding the number the TxPool is that you can't duplicate submission strategies.

I also think each `Submitter` should have its own TxPool.

> I aimed to leave a bit more flexibility in case we need each Submitter to also take an occasional look at other subpools

If we have `Arc<Mutex<TxPool>>` passed to each submitter, we can have a
`Vec<Arc<Mutex<TxPool>>>` held by the parent that it can pass around for sharing.

Reply:
> The issue with hardcoding the number the TxPool is that you can't duplicate submission strategies.

Yes, we would need to make a code change in that case: 
```
struct TxPool {
    eden: SubTxPool,
    flashbots: SubTxPool,
    flashbots_fast: SubTxPool,
    custom_nodes: SubTxPool,
}
```
But this can be solved with: https://github.com/cowprotocol/services/pull/235#discussion_r886526000

> I also think each Submitter should have its own TxPool.

Each Submitter has it's own SubTxPool?

> If we have Arc<Mutex<TxPool>> passed to each submitter, we can have a
> Vec<Arc<Mutex<TxPool>>> held by the parent that it can pass around for sharing.

I assume you wanted to write here: `Vec<Arc<Mutex<SubTxPool>>>`? I explicitly wanted to have a lock on a `TxPool` level rather then on a `SubTxPool` level because of the possible race conditions if `Submitter` wants to read other submitters pools...